internals/daemon/api_files.go

```diff
@@ -501,17 +501,53 @@ func writeFile(item writeFilesItem, source io.Reader) error {
 
 func mkdirAllUserGroup(path string, perm os.FileMode, uid, gid *int) error {
 	if uid != nil && gid != nil {
-		return mkdirAllChown(path, perm, sys.UserID(*uid), sys.GroupID(*gid))
+		return mkdir(
+			path,
+			perm,
+			&osutil.MkdirOptions{
+				MakeParents: true,
+				ExistOK:     true,
+				Chown:       true,
+				Chmod:       true,
+				UserID:      sys.UserID(*uid),
+				GroupID:     sys.GroupID(*gid),
+			},
+		)
 	} else {
-		return mkdirAllChown(path, perm, osutil.NoChown, osutil.NoChown)
+		return mkdir(
+			path,
+			perm,
+			&osutil.MkdirOptions{
+				MakeParents: true,
+				ExistOK:     true,
+				Chmod:       true,
+			},
+		)
 	}
 }
 
 func mkdirUserGroup(path string, perm os.FileMode, uid, gid *int) error {
```

Comment:
I think we should keep `mkdirAllUserGroup` because it's used twice. If we remove it, we have to copy-paste the same content to two functions.

`mkdirUserGroup` is only used once （in `makeDir` of the same file), we could remove it, but it's used in a function which is already over 20 lines of code, and has multiple if-elses. Removing it would increase the length of that function significantly and increase the cyclomatic complexity as well.

I prefer to keep them both.